x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -282,6 +284,61 @@ private static Map<String, Object> maybeRewriteMetadata(TransportVersion olderVe
         }
     }
 
+    /**
+     * Creates a copy of this Authentication instance, but only with metadata entries specified by `fieldsToKeep`.
+     * All other entries are removed from the copy's metadata.
+     */
+    public Authentication copyWithFilteredMetadataFields(final Set<String> fieldsToKeep) {
+        Objects.requireNonNull(fieldsToKeep);
+        if (fieldsToKeep.isEmpty()) {
+            return copyWithEmptyMetadata();
+        }
+        final Map<String, Object> metadataCopy = new HashMap<>(getAuthenticatingSubject().getMetadata());
+        final boolean metadataChanged = metadataCopy.keySet().retainAll(fieldsToKeep);
```

Comment:
Decided against an idempotency check (i.e., don't copy if it's unnecessary) here to keep the contract of the method clear. Callers can rely on always getting a new instance this way and don't have to worry about side-effects on the original instance. The idempotency check would only avoid the creation of a fairly small object, so not much performance gain.